internal/resolver/delegatingresolver/delegatingresolver.go

```diff
@@ -202,6 +209,31 @@ func needsProxyResolver(state *resolver.State) bool {
 	return false
 }
 
+func maybeAddDefaultPort(target, defaultPort string) (string, error) {
+	if target == "" {
+		return "", fmt.Errorf("missing address")
+	}
+	if host, port, err := net.SplitHostPort(target); err == nil {
+		if port == "" {
+			// If the port field is empty (target ends with colon), e.g. "[::1]:",
+			// this is an error.
+			return "", fmt.Errorf("missing port after port-separator colon")
+		}
+		// target has port, i.e ipv4-host:port, [ipv6-host]:port, host-name:port
+		if host == "" {
+			// Keep consistent with net.Dial(): If the host is empty, as in ":80",
+			// the local system is assumed.
+			host = "localhost"
+		}
+		return fmt.Sprintf("%s:%s", host, port), nil
```

Comment:
We should use [net.JoinHostPort](https://pkg.go.dev/net#JoinHostPort) to properly format this string. In case of IPv6 addresses for example, the resulting string may be `[%s]:%s` instead.